tools/umbra_layout.py

```diff
@@ -0,0 +1,566 @@
+            # We consider SAME as scale up.
+            return True
+        elif (scale == Scale.UP_2X):
+            return True
+        elif (scale == Scale.UP_4X):
+            return True
+        elif (scale == Scale.UP_8X):
+            return True
+        else:
+            sys.exit('Fatal Error: Unknown scale.')
+
+    def get_scale(scale):
+        # Returns whether the scale's value.
+        if (scale == Scale.DOWN_64X):
+            return 6
+        elif (scale == Scale.DOWN_32X):
+            return 5
+        elif (scale == Scale.DOWN_8X):
+            return 3
+        elif (scale == Scale.DOWN_4X):
+            return 2
+        elif (scale == Scale.DOWN_2X):
+            return 1
+        elif (scale == Scale.SAME):
+            return 0
+        elif (scale == Scale.UP_2X):
+            return 1
+        elif (scale == Scale.UP_4X):
+            return 2
+        elif (scale == Scale.UP_8X):
+            return 3
+        else:
+            sys.exit('Fatal Error: Unknown Scale.')
+
+
+class Region:
+    def __init__(self, desc, start, end):
+        # Description of region.
+        self.desc = desc
+        # Start of region.
+        self.start = start
+        # End of region.
+        self.end = end
+
+    def __str__(self):
+        formatted_start = get_formatted_hex(self.start)
+        formatted_end = get_formatted_hex(self.end)
+
+        # Return range as a string.
+        return '[' + formatted_start + ', ' + formatted_end + ']'
+
+    def __repr__(self):
+        return self.__str__()
+
+
+class RegionExpressionInfo:
+    def __init__(self, start_expr, end_expr):
+        # Expression of the start of the range.
+        self.start_expr = start_expr
+        # Expression of the end of the range.
+        self.end_expr = end_expr
+
+
+class Layout:
+    def __init__(self, mask, disp, unit, is_scale_up, scale, map_count):
+        # Mask used for shadow translation.
+        self.mask = mask
+        # Displacement used for shadow translation.
+        self.disp = disp
+        # Size of a unit.
+        self.unit = unit
+        # Denotes whether the shadow memory is scaled up or down.
+        self.is_scale_up = is_scale_up
+        # Denotes the scale of shadow memory.
+        self.scale = scale
+        # The maximum number of multiple maps supported by the layout.
+        self.map_count = map_count
+
+    def __translate_boundary(self, addr, map_index):
+        # Step 1: mask the app addr.
+        masked_addr = (addr & self.mask)
+
+        # Step 2: Derive the unit displacement.
+        unit_disp = self.unit * 2 * map_index
+        if self.is_scale_up:
+            unit_disp = unit_disp >> self.scale
+        else:
+            unit_disp = unit_disp << self.scale
+
+        # Step 3: Calculate the shadow address
+        shdw_addr = masked_addr + self.disp
+
+        # Step 4: Handle special case where top of addr is masked out, e.g., 0x800000000.
+        if addr != 0 and shdw_addr == self.disp:
```

Comment:
Not understanding this: this step would not be there for current umbra code??  Could you elaborate on this?

Reply:
That models this piece of code:

https://github.com/DynamoRIO/drmemory/blob/4fdcb814500a62317bf5659bba6a05a900c01dc9/umbra/umbra_64.c#L424

Reply:
Woah, still not understanding: inlined shadow translation code certainly does not have a branch and so this differs from what's inlined, so I'm not sure how it can work?  @zhaoqin do you recognize this?

Reply:
Yea.... I agree with your reading regarding inlining.

Basically, what the code is doing here is handling the case where the mask clears the top part of an address, which is essentially the data entropy required to set up a layout. 

if this condition, namely addr == map->disp, is confusing you, take a look at the previous line, where the addr is computed. If the mask zeros out the pc, then addr would be equal to map->disp.

https://github.com/DynamoRIO/drmemory/blob/4fdcb814500a62317bf5659bba6a05a900c01dc9/umbra/umbra_64.c#L423

Reply:
> If the mask zeros out the pc, then addr would be equal to map->disp.

Right, but that is expected: that's just part of the "mask and add" approach that you have to take into account when setting up the masks and disps.  If the pc is `0x800'00000001` we get disp+1 for addr; if the pc is one less we get disp.  Not sure *why* it tries to handle just the equals-disp part here and why it's so different from other cases.

The only thing I can think of is that it's trying to handle open endpoints.  So if this is our app region and it does not include 0x800':
```
 *   app2: [0x000007F0'00000000, 0x00000800'00000000): lib
```

Is it trying to handle someone iterating over that range and (incorrectly) *including* the endpoint or something, like someone printing out the endpoint of the corresponding shadow region?  This is something I've run into actually: line 51 at https://github.com/DynamoRIO/drmemory/pull/2301/files#diff-e88a4a2cc5d58ef58b542c5ec5473ac798f340de617d3b32bfc157fec6494566 where I subtract one from the end before translating to shadow, and then add 1 back on.

Reply:
Hmm, I think you are on to something there. That could explain why this special case is not handled in inlined code.

Should I, therefore, remove this special case from umbra_layout.py to reflect inlined code?

Reply:
Yes, I think it should be removed from the .py, and also removed from here with instead a tweak to whatever callers are trying to find shadow endpoints to have them do the -1,+1 solution -- but since that takes some work, esp to verify with the lack of good umbra tests, maybe in this PR just add a comment to the umbra_64.c code?

Reply:
I talked to @zhaoqin and he thinks it is indeed for the endpoint of a shadow block, and the -1,+1 solution in the callers is the alternative.

Reply:
Awesome, thanks for checking!

Reply:
I filed #2330 on the cleanup in the umbra code.

Reply:
Maybe as part of this PR you could put in a TODO i#2330 comment in the umbra code.

Reply:
Done.

Reply:
Let me know if this is okay. Hit resolve if it is fine.

Reply:
LGTM.